akka-docs/src/main/paradox/typed/reliable-delivery.md

```diff
@@ -0,0 +1,358 @@
+
+The producer and `WorkPullingProducerController` actors are supposed to be local so that these messages are
+fast and not lost.
+
+Similarly, on the consumer side the destination consumer actor will start the flow by sending an
+initial `ConsumerController.Start` message to the `ConsumerController`.
+
+Received messages from the producer are wrapped in `ConsumerController.Delivery` when sent to the consumer,
+which is supposed to reply with `ConsumerController.Confirmed` when it has processed the message.
+Next message is not delivered until the previous is confirmed. More messages from the producer that arrive
+while waiting for the confirmation are stashed by the `ConsumerController` and delivered when previous
+message was confirmed.
+
+The consumer and the `ConsumerController` actors are supposed to be local so that these messages are fast
+and not lost.
+
+Many unconfirmed messages can be in flight between the `WorkPullingProducerController` and each
+`ConsumerController`, but it is limited by a flow control window. The flow control is driven by the
+consumer side, which means that the `WorkPullingProducerController` will not send faster than the
+demand requested by the workers.
+
+Example of image converter worker (consumer):
+
+Scala
+:  @@snip [WorkPullingDocExample.scala](/akka-cluster-sharding-typed/src/test/scala/docs/delivery/WorkPullingDocExample.scala) { #imports #consumer }
+
+and image converter job manager (producer):
+
+Scala
+:  @@snip [WorkPullingDocExample.scala](/akka-cluster-sharding-typed/src/test/scala/docs/delivery/WorkPullingDocExample.scala) { #imports #producer }
+
+FIXME Java example
+
+Note how the `ActorRef` in the `Start` messages are constructed as message adapters to map the
+`RequestNext` and `Delivery` to the protocol of the producer and consumer actors respectively.
+
+### Work pulling delivery semantics
+
+For work pulling the order of the messages should not matter, because each message is routed randomly
+to one of the workers with demand.
+
+As long as neither producers nor workers crash (or workers being removed for other reasons) the messages are
+delivered to the workers without loss or duplicates. Meaning effectively once processing without any
+business level deduplication.
+
+Unconfirmed messages may be lost if the producer crashes. To avoid that you need to enable the @ref:[durable
+queue](#durable-producer) on the producer side. The stored unconfirmed messages will be redelivered when the
+corresponding producer is started again. Those messages may be routed to different workers than before
+and some of them may have already been processed but the fact that they were confirmed had not been stored
+yet. Meaning at-least once delivery.
+
+If a worker crashes or is stopped gracefully the unconfirmed messages will be redelivered to other workers.
+In that case some of these may already have been processed by the previous worker. Meaning at-least once delivery.
```

Comment:
Not sure I understand. The previous section was describing producer crashes and this section is about worker (consumer) crashes. Both may result in at-least once.

Reply:
Yes, but it still repeating the same. 

On the previous paragraph, we explain that messages may be redelivered because the confirmation is lost which give us at-least-once semantics. 

Then we kind of explain the same but in the context of crashing workers. 

The bottom line is: confirmation messages can be lost because the producer crashed before it receives it or worker crashes after processing and before confirming. In both cases, the durable queue on the producer side will ensure that the message is re-delivered to a worker (not necessarily the same as before) giving us at-least-once semantics. 

Anyway, I think we can re-write these two paragraphs by focusing on the fact that the durable queue gives at-least-once semantics and then explain the two cases in which that can happen.

Reply:
I think this is just a matter of taste of how to phrase it. I wrote it based on two scenarios, a) producer crashes, b) consumer crashes. In such semantical description I prefer to repeat some things to make it clear, rather than having a nice prose. I'm leaving this as is for now, but you are welcome to open a separate PR with suggested changes if you think it's important.